### PR TITLE
Add favorites button

### PR DESCRIPTION
**Pull request by kbs5280**
The button toggles between Add and Remove Favorite.
Clicking the button changes the state of the job listing item.

All updates to the code base are in the JobListItem.js file. Added the isFavorite state to the object with a default value of false. In the component's render section the `renderFavoritesButton` function is called. The function checks the state of the listing and displays the appropriate button, either `Add To Favorites` or `Remove From Favorites`. There is an onClick event set of the button, when the button is clicked the state is changed.

This feature is not tested in the application. Functionality can be observed in the browser.

We would like to add something to this feature like adding the favorites to localStorage and a search for Favorites.

### CODE REVIEW

File: lib/components/JobListItem.js [hunk 1]
```diff
@@ -36,6 +39,29 @@ class JobListItem extends React.Component {
...
+    if (!this.state.isFavorite) {
+      return (
+        <div>
+          <button className="btn btn-default" onClick={this.addToFavorites.bind(this)}>Add To Favorites</button>
+        </div>
+      )
+    }
+    return (
+      <div>
+        <button className="btn btn-default" onClick={this.removeFromFavorites.bind(this)}>Remove From Favorites</button>
```
Comment (pwentz) [inline lib/components/JobListItem.js]: nice use of `.bind()` here. This is a perfectly viable option, but if you ever get tired of adding `.bind(this)` to every event callback an alternative option would be to bind the function just once in your constructor:
```
constructor() {
  super()
  this.removeFromFavorites = this.removeFromFavorites.bind(this)
  this.addToFavorites = this.addToFavorites.bind(this)
  this.state = { isFavorite: false }
```

File: lib/components/JobListItem.js [hunk 2]
```diff
@@ -36,6 +39,29 @@ class JobListItem extends React.Component {
...
+      </div>
+    )
+  }
+
+  addToFavorites() {
+    this.setState({ isFavorite: true });
+  }
+
+  removeFromFavorites() {
+    this.setState({ isFavorite: false });
```
Comment (pwentz) [inline lib/components/JobListItem.js]: You can create toggle functionality (if that's what you're looking for) by doing something like:
```
toggleFavorite() {
  this.setState({ isFavorite: !this.state.isFavorite })
}
```

File: lib/components/JobListItem.js [hunk 3]
```diff
@@ -36,6 +39,29 @@ class JobListItem extends React.Component {
       return <p>Loading...</p>
     }
   }
+
+  renderFavoritesButton() {
+    if (!this.state.isFavorite) {
```
Comment (wfro) [inline lib/components/JobListItem.js]: Good use of a guard clause here and no else after return  👍 , both of those are popular React conventions these days.

File: lib/components/JobListItem.js (hunk 1 above)
Comment (kbs5280) [inline lib/components/JobListItem.js]: @pwentz  and @wrfo That's very helpful and informative. I am going to change this.  Thanks for the explanation of what this is better and more performant.

File: lib/components/JobListItem.js (hunk 2 above)
Comment (kbs5280) [inline lib/components/JobListItem.js]: Great point. I will update this.

File: lib/components/JobListItem.js (hunk 3 above)
Comment (kbs5280) [inline lib/components/JobListItem.js]: Thanks!

**Comment by zackforbing** [comment]
after our work on this today, I can attest to its functionality. merging.

**Pull request merged**